File: plugins/oracle/scripts/mcp-fleet/services/notion.mjs
// Notion workspace probe.
//
// @notionhq/notion-mcp-server consumes an integration token via the
// NOTION_TOKEN env. Each Notion workspace requires its own internal
// integration; the secret_xxx token is workspace-scoped, so multi-workspace
// = multiple integrations.
//
// Auto-extraction of the secret is not feasible: Notion's integration
// secret is revealed via a Show button and the value is rendered into a
// React-controlled <input> with no stable selector across UI revs. Manual
// paste is the honest path.

import { launchProfile } from "../lib/playwright-launcher.mjs";
import { ask, askSecret } from "../lib/prompt.mjs";
import { upsert } from "../lib/workspaces-store.mjs";

export async function probe({ label }) {
  const wsLabel = label || (await ask("Notion workspace label (e.g. 'acme'): "));
  if (!wsLabel) throw new Error("workspace label is required");

  process.stdout.write(
    `\nLaunching Chrome with an isolated profile for notion/${wsLabel}.\n` +
    `Sign in to the target workspace, then create an internal integration:\n` +
    `  1. Go to https://www.notion.so/profile/integrations\n` +
    `  2. New integration, name it (the workspace dropdown picks the binding)\n` +
    `  3. Copy the Internal Integration Secret (starts with secret_ or ntn_)\n` +
    `  4. Share at least one page/database with the integration so it has data scope\n\n`,
  );

  const { close } = await launchProfile({
    service: "notion",
    workspace: wsLabel,
    startUrl: "https://www.notion.so/profile/integrations",
  });

  const token = await askSecret("Paste the integration token (input hidden): ");
  await close();

  if (!token || !(token.startsWith("secret_") || token.startsWith("ntn_"))) {
    throw new Error("Expected a Notion integration token starting with secret_ or ntn_");
  }

  return upsert({
    service: "notion",
    label: wsLabel,
    kind: "notion-integration",
    credentials: { NOTION_TOKEN: token },
    profileDir: "",
  });
}

if (import.meta.url === `file://${process.argv[1]}`) {
  probe({ label: process.argv[2] }).then(
    (e) => { process.stdout.write(`Persisted notion workspace ${e.label}\n`); process.exit(0); },
    (e) => { process.stderr.write(`${e.message}\n`); process.exit(1); },
  );
}

File: plugins/oracle/scripts/mcp-fleet-py/mcp_fleet/services/notion.py
"""Notion workspace probe.

@notionhq/notion-mcp-server consumes NOTION_TOKEN. Each Notion workspace
requires its own internal integration; the integration secret is
workspace-scoped, so multi-workspace = multiple integrations.

After paste we GET /v1/users/me to validate and discover the workspace
name. Important warning printed: if no page/db is shared with the
integration, the MCP server will see an empty workspace.
"""

from __future__ import annotations

import httpx

from .. import store
from ..paths import profile_dir
from .base import ProbeContext, ProbeError


NOTION_API_VERSION = "2022-06-28"


async def _validate(token: str) -> tuple[str, str]:
    """Returns (workspace_name, bot_name)."""
    async with httpx.AsyncClient(timeout=15) as client:
        r = await client.get(
            "https://api.notion.com/v1/users/me",
            headers={
                "Authorization": f"Bearer {token}",
                "Notion-Version": NOTION_API_VERSION,
            },
        )
    if r.status_code != 200:
        raise ProbeError(f"Notion API rejected the token ({r.status_code}): {r.text[:200]}")
    body = r.json()
    bot = body.get("bot") or {}
    ws_name = bot.get("workspace_name") or "Notion workspace"
    bot_name = body.get("name") or "integration"
    return ws_name, bot_name


async def probe(label: str, ctx: ProbeContext) -> store.Workspace:
    if not label:
        raise ProbeError("workspace label is required")

    ctx.log(f"opening Chromium at notion.so/profile/integrations for notion/{label}")
    from .browser_helper import open_url_in_profile
    async with open_url_in_profile(
        service="notion",
        workspace=label,
        url="https://www.notion.so/profile/integrations",
    ):
        ctx.log("create new integration; the workspace dropdown picks the binding")
        ctx.log("after creation, copy the Internal Integration Secret")
        token = (await ctx.paste_secret(
            "Paste the integration token (starts with secret_ or ntn_)"
        )).strip()

    if not (token.startswith("secret_") or token.startswith("ntn_")):
        raise ProbeError("expected secret_ or ntn_ prefix")

    ctx.log("validating against api.notion.com ...")
    ws_name, bot_name = await _validate(token)
    ctx.log(f"validated -- workspace: {ws_name} (bot: {bot_name})")
    ctx.log("REMINDER: share at least one page/db with the integration in Notion's UI,")
    ctx.log("otherwise the MCP server will see an empty workspace.")

    return store.upsert(store.Workspace(
        service="notion",
        label=label,
        kind="notion-integration",
        credentials={"NOTION_TOKEN": token},
        profileDir=str(profile_dir("notion", label)),
        discoveredName=ws_name,
    ))
